File: backend/services/mesh/mesh_merkle.py
from __future__ import annotations

import hashlib
from typing import Any


def _hash_bytes(data: bytes) -> str:
    return hashlib.sha256(data).hexdigest()


def hash_leaf(value: str) -> str:
    return _hash_bytes(value.encode("utf-8"))


def hash_pair(left: str, right: str) -> str:
    return _hash_bytes(f"{left}{right}".encode("utf-8"))


def build_merkle_levels(leaves: list[str]) -> list[list[str]]:
    if not leaves:
        return []
    level = [hash_leaf(leaf) for leaf in leaves]
    levels = [level]
    while len(level) > 1:
        next_level: list[str] = []
        for idx in range(0, len(level), 2):
            left = level[idx]
            right = level[idx + 1] if idx + 1 < len(level) else left
            next_level.append(hash_pair(left, right))
        level = next_level
        levels.append(level)
    return levels


def merkle_root(leaves: list[str]) -> str:
    levels = build_merkle_levels(leaves)
    if not levels:
        return ""
    return levels[-1][0]


def merkle_proof_from_levels(levels: list[list[str]], index: int) -> list[dict[str, Any]]:
    if not levels:
        return []
    if index < 0 or index >= len(levels[0]):
        return []
    proof: list[dict[str, Any]] = []
    idx = index
    for level in levels[:-1]:
        is_right = idx % 2 == 1
        sibling_idx = idx - 1 if is_right else idx + 1
        if sibling_idx >= len(level):
            sibling_hash = level[idx]
        else:
            sibling_hash = level[sibling_idx]
        proof.append({"hash": sibling_hash, "side": "left" if is_right else "right"})
        idx //= 2
    return proof


def verify_merkle_proof(
    leaf_value: str, index: int, proof: list[dict[str, Any]], root: str
) -> bool:
    current = hash_leaf(leaf_value)
    idx = index
    for step in proof:
        sibling = str(step.get("hash", ""))
        side = str(step.get("side", "right")).lower()
        if side == "left":
            current = hash_pair(sibling, current)
        else:
            current = hash_pair(current, sibling)
        idx //= 2
    return current == root

File: frontend/src/mesh/meshMerkle.ts
type MerkleStep = { hash: string; side: 'left' | 'right' | string };

function bufToHex(buffer: ArrayBuffer): string {
  const bytes = new Uint8Array(buffer);
  let out = '';
  for (let i = 0; i < bytes.length; i += 1) {
    out += bytes[i].toString(16).padStart(2, '0');
  }
  return out;
}

async function sha256Hex(data: string): Promise<string> {
  if (typeof crypto !== 'undefined' && crypto.subtle) {
    const buf = await crypto.subtle.digest('SHA-256', new TextEncoder().encode(data));
    return bufToHex(buf);
  }
  const mod = await import('crypto');
  return mod.createHash('sha256').update(data).digest('hex');
}

async function hashLeaf(value: string): Promise<string> {
  return sha256Hex(value);
}

async function hashPair(left: string, right: string): Promise<string> {
  return sha256Hex(`${left}${right}`);
}

export async function buildMerkleRoot(leaves: string[]): Promise<string> {
  if (!leaves.length) return '';
  let level = await Promise.all(leaves.map((leaf) => hashLeaf(leaf)));
  while (level.length > 1) {
    const next: string[] = [];
    for (let i = 0; i < level.length; i += 2) {
      const left = level[i];
      const right = level[i + 1] ?? left;
      next.push(await hashPair(left, right));
    }
    level = next;
  }
  return level[0];
}

export async function verifyMerkleProof(
  leafValue: string,
  index: number,
  proof: MerkleStep[],
  root: string,
): Promise<boolean> {
  let current = await hashLeaf(leafValue);
  let idx = index;
  for (const step of proof) {
    const sibling = step.hash ?? '';
    const side = String(step.side || 'right').toLowerCase();
    if (side === 'left') {
      current = await hashPair(sibling, current);
    } else {
      current = await hashPair(current, sibling);
    }
    idx = Math.floor(idx / 2);
  }
  return current === root;
}

export type { MerkleStep };
